### PR TITLE
Adds parameter information to ApiExplorer

### PR DESCRIPTION
**Pull request by rynowak**
This change makes ApiDescription and ApiParameterDescription aware of all
of the new features we built into model binding for enhanced DTO support
(uber-binding).

The main change is that instead of sticking just to the declared
parameters on the action itself, we now traverse model metadata and break
the parameters down based on their logical data source.

This means that a model like the below will yield 3 parameters:

```
public class ProductChangeCommandDTO
{
    public int Id { get; set; }

    [FromBody]
    public ProductDetails Changes { get; set; }

    [FromQuery]
    public string AdminComments { get; set; }

    [FromServices]
    public IProductRepository Repository { get; set; }
}
```

The 'Repository' will be hidden, as it's not related to user input.

Additionally, we treat different sources differently. In the
above example, 'Changes' is from the body and will be treated as a
leaf-node.

However if you use nested DTOs that are bound from the query string (using
`[FromQuery]`) or similar, we'll recursively explore to find as much
structure as possible.

This information is combined with data from the route template to give a
much more complete picture than we ever could in the past for parameters,
especially when DTO/Command pattern is used.

### CODE REVIEW

**Comment by rynowak** [comment]
**Note that this PR has the design tag**

What's missing:
- Doc comments
- Functional tests (current tests are passing, need to add more)
- Updates to the sample

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs [hunk 1]
```diff
@@ -450,5 +385,279 @@ private IApiResponseMetadataProvider[] GetResponseMetadataAttributes(ControllerA
...
+                    // Name was explicitly provided
+                    return metadata.BinderModelName;
+                }
+                else
+                {
+                    return ModelBindingHelper.CreatePropertyModelName(containerName, metadata.PropertyName);
+                }
+            }
+
+            // This isn't extensible right now.
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: `TryGetSource` is intentionally not extensible right now. We might want to add something like this in the future based on demand, but for now, our knowledge of what sources exist is hardcoded.

File: src/Microsoft.AspNet.Mvc.Core/ParameterBinding/ModelBindingHelper.cs [hunk 2]
```diff
@@ -246,7 +246,7 @@ internal static string GetPropertyName(Expression expression)
                  StringComparison.OrdinalIgnoreCase);
         }
 
-        private static string CreatePropertyModelName(string prefix, string propertyName)
+        public static string CreatePropertyModelName(string prefix, string propertyName)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ParameterBinding/ModelBindingHelper.cs]: Need to use this in api explorer to build the parameter names

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/ICompositeModelBinder.cs [hunk 3]
```diff
@@ -8,5 +10,6 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
     /// </summary>
     public interface ICompositeModelBinder : IModelBinder
     {
+        IReadOnlyList<IModelBinder> ModelBinders { get; }
     }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/ICompositeModelBinder.cs]: will revert these changes

File: test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs [hunk 4]
```diff
@@ -110,49 +111,6 @@ public void GetApiDescription_CreatesMultipleDescriptionsForMultipleHttpMethods(
             Assert.Single(descriptions, d => d.HttpMethod == "GET");
         }
 
-        // This is a test for the placeholder behavior - see #886
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs]: **RIP in PEACE**
Placeholder Test
2014-2015 

we barely knew thee

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs [hunk 5]
```diff
@@ -450,5 +385,279 @@ private IApiResponseMetadataProvider[] GetResponseMetadataAttributes(ControllerA
...
+                    source = null;
+                    return false;
+                }
+
+                if (metadata.BinderMetadata is IFormatterBinderMetadata)
+                {
+                    source = ApiParameterSource.Body;
+                    return true;
+                }
+                else if (metadata.BinderMetadata is IHeaderBinderMetadata)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: nit: drop the else ..

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs [hunk 6]
```diff
@@ -450,5 +385,279 @@ private IApiResponseMetadataProvider[] GetResponseMetadataAttributes(ControllerA
...
+                {
+                    return ModelBindingHelper.CreatePropertyModelName(containerName, metadata.PropertyName);
+                }
+            }
+
+            // This isn't extensible right now.
+            //
+            // Returns true if the source is greedy (means to stop exploring the model)
+            // Returns false if the source in unknown or known but not greedy (like [FromQuery])
+            private static bool TryGetSource(ModelMetadata metadata, out ApiParameterSource source)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: Lets rename it . 
In a typical Try* pattern we would expect the out parameter to be null when Try* returns false.

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs [hunk 7]
```diff
@@ -450,5 +385,279 @@ private IApiResponseMetadataProvider[] GetResponseMetadataAttributes(ControllerA
...
+                    Name = GetName(containerName, metadata),
+                    ParameterDescriptor = Parameter,
+                    Source = source,
+                    Type = metadata.ModelType,
+                };
+            }
+
+            private static string GetName(string containerName, ModelMetadata metadata)
+            {
+                if (!String.IsNullOrEmpty(metadata.BinderModelName))
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: ```string```

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs [hunk 8]
```diff
@@ -450,5 +385,279 @@ private IApiResponseMetadataProvider[] GetResponseMetadataAttributes(ControllerA
...
+            private HashSet<PropertyKey> Visited { get; }
+
+            public void WalkParameter()
+            {
+                var modelMetadata = Context.MetadataProvider.GetMetadataForParameter(
+                    modelAccessor: null,
+                    methodInfo: Context.ActionDescriptor.MethodInfo,
+                    parameterName: Parameter.Name);
+
+                var binderMetadata = Parameter.BinderMetadata;
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: move this inside if.

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs [hunk 9]
```diff
@@ -450,5 +385,279 @@ private IApiResponseMetadataProvider[] GetResponseMetadataAttributes(ControllerA
...
+                    // If we get here, then it means we didn't find a match for any of the model. This means that it's
+                    // likely 'model-bound' in the traditional MVC sense (formdata + query string + route data) and
+                    // doesn't use any IBinderMetadata.
+                    // 
+                    // Add a single 'default' parameter description for the model.
+                    Context.Results.Add(CreateResult(modelMetadata, source, containerName: string.Empty));
+                }
+            }
+
+            private bool Visit(ModelMetadata modelMetadata, ApiParameterSource ambientSource, string containerName)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: some comments on when this is supposed to return true/false would be helpful.

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs [hunk 10]
```diff
@@ -450,5 +385,279 @@ private IApiResponseMetadataProvider[] GetResponseMetadataAttributes(ControllerA
...
+
+                    return true;
+                }
+             
+                if (!modelMetadata.Properties.Any())
+                {
+                    if (source == null)
+                    {
+                        // If there are no properties and no source then we don't know how to bind this.
+                        // Return and let it get potentially merged with the container.
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: Sorry did not understand ```merged with the container.```

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs [hunk 11]
```diff
@@ -450,5 +385,279 @@ private IApiResponseMetadataProvider[] GetResponseMetadataAttributes(ControllerA
...
+                // This should result in two parameters:
+                //
+                //  AccountId - source: Any
+                //  Order - source: Body
+                //
+
+                var propertyCount = 0;
+                var unboundProperties = new HashSet<ModelMetadata>();
+
+                // We don't want to append the parameter name to the model name when recursing.
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: Is this correct? wouldnt the containerType be non null when visiting inner properties?

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: nah

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: sure, changing it to just be `GetSource`. I don't want to fixate on this because it's going to change substantially in the next iteration.

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs (hunk 8 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: the `if` got removed, it shouldn't have been there anyway.

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs (hunk 9 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: added some comments

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs (hunk 10 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: reworded a little, let me know if you like it better. without a suggestion I can't really improve it more sadly.

File: src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs (hunk 11 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Description/DefaultApiDescriptionProvider.cs]: `containerType` will be null iff we're visiting the top-level object.

**Comment by rynowak** [comment]
@harshgMSFT updated - now with functional tests.

Note that there are already really complete tests for the route-info part of the parameter data. I had to update a few things based on our decisions today.

File: test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs [hunk 12]
```diff
@@ -584,6 +523,360 @@ public void GetApiDescription_IncludesResponseFormats_FilteredByType()
...
+            // Arrange
+            var action = CreateActionDescriptor(nameof(AcceptsCycle));
+            var parameterDescriptor = action.Parameters.Single();
+
+            // Act
+            var descriptions = GetApiDescriptions(action);
+
+            // Assert
+            var description = Assert.Single(descriptions);
+            Assert.Equal(1, description.ParameterDescriptions.Count);
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs]: nit: not needed? we do a single on the next line.

File: test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs [hunk 13]
```diff
@@ -584,6 +523,360 @@ public void GetApiDescription_IncludesResponseFormats_FilteredByType()
...
+
+            var c = Assert.Single(description.ParameterDescriptions);
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs]: better name ?

File: test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs (hunk 12 above)
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs]: same comment for other places  in tests

File: test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs (hunk 13 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs]: All of these are named the same as the parameter/property that they represent. Suggestion?

File: test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs [hunk 14]
```diff
@@ -584,6 +523,360 @@ public void GetApiDescription_IncludesResponseFormats_FilteredByType()
...
+            Assert.Same(ApiParameterSource.ModelBinding, c.Source);
+            Assert.Equal("c", c.Name);
+            Assert.Equal(typeof(HasCollection_Complex), c.Type);
+        }
+
+        [Fact]
+        public void GetApiDescription_ParameterDescription_RedundentMetadataMergedWithParent()
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs]: what would be the result of 
```c# 
void Action([FromForm] Person p)
{
}

public class Person 
{
         [FromHeader("Name")]
         public string Name { get; set; }
         
         [FromForm]  
         public int Id { get; set; }
}
```
probably a test case would help.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ApiExplorerTest.cs [hunk 15]
```diff
@@ -722,6 +723,163 @@ public class ApiExplorerTest
...
+            Assert.Equal(ApiParameterSource.Query.Id, shippingInstructions.Source);
+            Assert.Equal(typeof(string).FullName, shippingInstructions.Type);
+
+            var feedback = Assert.Single(parameters, p => p.Name == "Comments.Feedback");
+            Assert.Equal(ApiParameterSource.Form.Id, feedback.Source);
+            Assert.Equal(typeof(string).FullName, feedback.Type);
+        }
+
+        [FromQuery]
+        public string ShippingInstructions { get; set; }
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ApiExplorerTest.cs]: is this used somewhere?

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ApiExplorerTest.cs [hunk 16]
```diff
@@ -722,6 +723,163 @@ public class ApiExplorerTest
...
+
+        public string Feedback { get; set; }
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ApiExplorerTest.cs]: used somewhere?

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataTest.cs [hunk 17]
```diff
@@ -151,6 +155,51 @@ public void IsComplexTypeTestsReturnsTrueForComplexTypes(Type type)
...
+        [InlineData(typeof(int))]
+        [InlineData(typeof(NonCollectionType))]
+        [InlineData(typeof(string))]
+        public void IsCollectionType_NonCollectionTypes(Type type)
+        {
+            // Arrange
+            var provider = new EmptyModelMetadataProvider();
+
+            // Act
+            var modelMetadata = new ModelMetadata(provider, null, null, type, null);
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataTest.cs]: named parameters please.

File: test/WebSites/ApiExplorerWebSite/ApiExplorerDataFilter.cs [hunk 18]
```diff
@@ -53,13 +59,21 @@ private ApiExplorerData CreateSerializableData(ApiDescription description)
...
-                    ConstraintTypes = parameter?.Constraints?.Select(c => c.GetType().Name).ToArray(),
+                    Source = parameter.Source.Id,
+                    Type = parameter.Type?.FullName,
                 };
 
+                if (parameter.RouteInfo != null)
+                {
+                    parameterData.RouteInfo = new ApiExplorerParameterRouteInfo()
+                    {
+                        ConstraintTypes = parameter.RouteInfo.Constraints?.Select(c => c.GetType().Name).ToArray(),
```
Comment (harshgMSFT) [inline test/WebSites/ApiExplorerWebSite/ApiExplorerDataFilter.cs]: nit: ```c```-> ```constraint```

File: test/WebSites/ApiExplorerWebSite/ApiExplorerDataFilter.cs [hunk 19]
```diff
@@ -9,11 +10,11 @@
 namespace ApiExplorerWebSite
 {
     /// <summary>
-    /// An action filter that looks up and serializes Api Explorer data for the action.
+    /// A resource filter that looks up and serializes Api Explorer data for the action.
     ///
     /// This replaces the 'actual' output of the action.
     /// </summary>
-    public class ApiExplorerDataFilter : ActionFilterAttribute
+    public class ApiExplorerDataFilter : IResourceFilter
```
Comment (harshgMSFT) [inline test/WebSites/ApiExplorerWebSite/ApiExplorerDataFilter.cs]: Wondering if there is some other reason for this other than  we want this to run before model binding?? (so that we can just consume it without actually invoking the action?)

File: src/Microsoft.AspNet.Mvc.Core/ControllerActionDescriptorBuilder.cs [hunk 20]
```diff
@@ -290,6 +290,13 @@ private static ParameterDescriptor CreateParameterDescriptor(ParameterModel para
             var apiExplorerIsVisible = action.ApiExplorer?.IsVisible ?? controller.ApiExplorer?.IsVisible ?? false;
             if (apiExplorerIsVisible)
             {
+                if (!IsAttributeRoutedAction(actionDescriptor))
+                {
+                    // ApiExplorer is only supported on attribute routed actions.
+                    throw new InvalidOperationException(Resources.FormatApiExplorer_UnsupportedAction(
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ControllerActionDescriptorBuilder.cs]: Add a test for exception.

File: test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs (hunk 13 above)
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs]: yeah i realized that, probably rename the parameter/property and have the same name here ? 
for example ```AcceptsHasCollection(HasCollection c)``` to ```AcceptsHasCollection(HasCollection collection)``` ...
consider this as a nit.

**Comment by harshgMSFT** [comment]
Looks close to done .. :watch:  on couple of tests and minor comments.

File: test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs (hunk 12 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs]: Fixed all of these

File: test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs (hunk 14 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Description/DefaultApiDescriptionProviderTest.cs]: Two parameters. We never partially describe a model. Adding a test for this case.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ApiExplorerTest.cs (hunk 15 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ApiExplorerTest.cs]: removed

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ApiExplorerTest.cs (hunk 16 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ApiExplorerTest.cs]: removed

File: test/WebSites/ApiExplorerWebSite/ApiExplorerDataFilter.cs (hunk 18 above)
Comment (rynowak) [inline test/WebSites/ApiExplorerWebSite/ApiExplorerDataFilter.cs]: nope

File: test/WebSites/ApiExplorerWebSite/ApiExplorerDataFilter.cs (hunk 19 above)
Comment (rynowak) [inline test/WebSites/ApiExplorerWebSite/ApiExplorerDataFilter.cs]: That's exactly the reason. No reason to actually run model binding, that just makes it hard to author the tests.

**Comment by rynowak** [comment]
Already added `ApiExplorer_ThrowsForContentionalRouting`

**Comment by harshgMSFT** [comment]
my bad .. didn't see that somehow, :shipit: after  fixing minor comments in tests.

**Comment by rynowak** [comment]
pushed an update addressing minor stuff. Going to rebase on release and push there
